File: curve.c
#include "curve.h"

#define SC(x) static const Curve static_##x = { \
	.name = L"Fixed " L#x L"%", \
	.length = 1, \
	.items = { (x) } \
}
#define pSC(x) &static_##x

SC(25);
SC(50);
SC(75);
SC(100);

static const Curve preset_fan_silent = {
	.name = L"Silent",
	.length = 21,
	.items = { 25, 25, 25, 25, 25, 25, 25, 25, 35, 45, 55, 75, 100, 100, 100, 100, 100, 100, 100, 100, 100 },
};

static const Curve preset_fan_performance = {
	.name = L"Performance",
	.length = 21,
	.items = { 50, 50, 50, 50, 50, 50, 50, 50, 60, 70, 80, 90, 100, 100, 100, 100, 100, 100, 100, 100, 100 },
};

static const Curve preset_pump_silent = {
	.name = L"Silent",
	.length = 21,
	.items = { 60, 60, 60, 60, 60, 60, 60, 60, 70, 80, 90, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100 },
};

static const Curve preset_pump_performance = {
	.name = L"Performance",
	.length = 21,
	.items = { 70, 70, 70, 70, 70, 70, 70, 70, 80, 85, 90, 95, 100, 100, 100, 100, 100, 100, 100, 100, 100 },
};

const Curve *Curve_fan_presets[] = {
	&preset_fan_silent,
	&preset_fan_performance,
	pSC(25),
	pSC(50),
	pSC(75),
	pSC(100),
	NULL,
};

const Curve *Curve_pump_presets[] = {
	&preset_pump_silent,
	&preset_pump_performance,
	pSC(25),
	pSC(50),
	pSC(75),
	pSC(100),
	NULL,
};

File: curve.zig
const std = @import("std");

pub const Curve = extern struct {
    name: [16:0]u16,
    length: u8,
    items: [31]u8,

    fn fromSlice(comptime name: [:0]const u8, items: []const u8) Curve {
        var c: Curve = undefined;
        const u16name = std.unicode.utf8ToUtf16LeStringLiteral(name);
        @memcpy(c.name[0..u16name.len], u16name);
        c.length = items.len;
        @memcpy(c.items[0..c.length], items);
        return c;
    }

    fn fromFixed(pct: u8) Curve {
        var c: Curve = undefined;
        const u8name = std.fmt.comptimePrint("Fixed {d}%", .{pct});
        const name = std.unicode.utf8ToUtf16LeStringLiteral(u8name);
        @memcpy(c.name[0..name.len], name);
        c.length = 1;
        c.items[0] = pct;
        return c;
    }

    pub fn toSlice(c: *const Curve) []const u8 {
        return c.items[0..c.length];
    }
};

const presets = struct {
    const fan: [6:null]?*const Curve = .{
        &.fromSlice("Silent", &.{ 25, 25, 25, 25, 25, 25, 25, 25, 35, 45, 55, 75, 100, 100, 100, 100, 100, 100, 100, 100, 100 }),
        &.fromSlice("Performance", &.{ 50, 50, 50, 50, 50, 50, 50, 50, 60, 70, 80, 90, 100, 100, 100, 100, 100, 100, 100, 100, 100 }),
        &.fromFixed(25),
        &.fromFixed(50),
        &.fromFixed(75),
        &.fromFixed(100),
    };

    const pump: [6:null]?*const Curve = .{
        &.fromSlice("Silent", &.{ 60, 60, 60, 60, 60, 60, 60, 60, 70, 80, 90, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100 }),
        &.fromSlice("Performance", &.{ 70, 70, 70, 70, 70, 70, 70, 70, 80, 85, 90, 95, 100, 100, 100, 100, 100, 100, 100, 100, 100 }),
        &.fromFixed(25),
        &.fromFixed(50),
        &.fromFixed(75),
        &.fromFixed(100),
    };
};

comptime {
    @export(&presets.fan, .{ .name = "Curve_fan_presets", .linkage = .strong });
    @export(&presets.pump, .{ .name = "Curve_pump_presets", .linkage = .strong });
}
